shell/renderer/api/atom_api_context_bridge.cc

```diff
@@ -0,0 +1,507 @@
+  mate::Dictionary api(source_context->GetIsolate(), api_object);
+  mate::Dictionary proxy =
+      mate::Dictionary::CreateEmpty(destination_context->GetIsolate());
+  store->CacheProxiedObject(api.GetHandle(), proxy.GetHandle());
+  auto maybe_keys = api.GetHandle()->GetOwnPropertyNames(
+      source_context,
+      static_cast<v8::PropertyFilter>(v8::ONLY_ENUMERABLE | v8::SKIP_SYMBOLS),
+      v8::KeyConversionMode::kConvertToString);
+  if (maybe_keys.IsEmpty())
+    return v8::MaybeLocal<v8::Object>(proxy.GetHandle());
+  auto keys = maybe_keys.ToLocalChecked();
+
+  v8::Context::Scope destination_context_scope(destination_context);
+  {
+    uint32_t length = keys->Length();
+    std::string key_str;
+    for (uint32_t i = 0; i < length; i++) {
+      v8::Local<v8::Value> key =
+          keys->Get(destination_context, i).ToLocalChecked();
+      // Try get the key as a string
+      if (!mate::ConvertFromV8(api.isolate(), key, &key_str)) {
+        continue;
+      }
+      v8::Local<v8::Value> value;
+      if (!api.Get(key_str, &value))
+        continue;
+
+      auto passed_value =
+          PassValueToOtherContext(source_context, destination_context, value,
+                                  store, recursion_depth + 1);
+      if (passed_value.IsEmpty())
+        return v8::MaybeLocal<v8::Object>();
+      proxy.Set(key_str, passed_value.ToLocalChecked());
+    }
+
+    return proxy.GetHandle();
+  }
+}
+
+#ifdef DCHECK_IS_ON
+mate::Dictionary DebugGC(mate::Dictionary empty) {
+  auto* render_frame = GetRenderFrame(empty.GetHandle());
+  auto* store = GetOrCreateStore(render_frame);
+  mate::Dictionary ret = mate::Dictionary::CreateEmpty(empty.isolate());
+  ret.Set("functionCount", store->functions().size());
+  auto* proxy_map = store->proxy_map();
+  ret.Set("objectCount", proxy_map->size() * 2);
+  int live_from = 0;
+  int live_proxy = 0;
+  for (auto iter = proxy_map->begin(); iter != proxy_map->end(); iter++) {
+    auto* node = iter->second;
+    while (node) {
+      if (!std::get<0>(node->pair).IsEmpty())
+        live_from++;
+      if (!std::get<1>(node->pair).IsEmpty())
+        live_proxy++;
+      node = node->next;
+    }
+  }
+  ret.Set("liveFromValues", live_from);
+  ret.Set("liveProxyValues", live_proxy);
+  return ret;
+}
+#endif
+
+void ExposeAPIInMainWorld(const std::string& key,
+                          v8::Local<v8::Object> api_object,
+                          mate::Arguments* args) {
+  auto* render_frame = GetRenderFrame(api_object);
+  CHECK(render_frame);
+  context_bridge::RenderFramePersistenceStore* store =
+      GetOrCreateStore(render_frame);
+  auto* frame = render_frame->GetWebFrame();
+  CHECK(frame);
+  v8::Local<v8::Context> main_context = frame->MainWorldScriptContext();
+  mate::Dictionary global(main_context->GetIsolate(), main_context->Global());
+
+  if (global.Has(key)) {
+    args->ThrowError(
+        "Cannot bind an API on top of an existing property on the window "
```

Comment:
Is there a technical reason not to do this or is it just sort of a bad idea? If a user wants to do this, they can easily do it:

```js
contextBridge.exposeInMainWorld('_myFakeLocation', {...})
webFrame.executeJavaScript('window.location = _myFakeLocation')
```

so it seems a bit weird to throw an error here.

Reply:
> Is there a technical reason not to do this or is it just sort of a bad idea?

Overriding some globals will throw errors and also it's kinda a bad idea.  Stopping users shooting themselves in the foot or at least making it harder was important here.

Reply:
Counterpoint: if a user exposes an API at a key that isn't currently present on `window`, but later becomes present due to a new Web API, their app will break for no good reason. I get the idea of trying to help users not shoot themselves in the foot, i'm just not sure the juice is worth the squeeze here.

Reply:
Further counter point, wouldn't you want to know if you were suddenly overwriting a new web API that any one of your thousands of dependencies could be using?